src/module/actor/character/elemental-blast.ts

```diff
@@ -0,0 +1,334 @@
+            const flag = item.flags.pf2e.damageSelections;
+            return isObject<Record<string, unknown>>(flag) ? flag : {};
+        })();
+        const blasts = Object.values(kineticist.elementalBlast).map((b) => schema.clean(b));
+
+        const validationFailures = blasts.flatMap((b) => schema.validate(b) ?? []);
+        for (const failure of validationFailures) {
+            throw failure.asError();
+        }
+
+        const domains = [...statistic.domains, ...statistic.check.domains, "elemental-blast-attack-roll"];
+        const maps = calculateMAPs(item, { domains, options: this.actor.getRollOptions(domains) });
+
+        return blasts.map((blast) => {
+            const damageTypes: BlastConfigDamageType[] = blast.damageTypes.map((dt) => ({
+                value: dt,
+                label: CONFIG.PF2E.damageTypes[dt],
+                selected: damageTypeSelections[blast.element] === dt,
+                glyph: DAMAGE_TYPE_ICONS[dt] ?? "",
+            }));
+            const firstDamageType = damageTypes.at(0);
+            if (firstDamageType && !damageTypes.some((dt) => dt.selected)) {
+                firstDamageType.selected = true;
+            }
+
+            return {
+                ...blast,
+                statistic,
+                map1: statistic.check.mod + maps.map1,
+                map2: statistic.check.mod + maps.map2,
+                item,
+                actionCost,
+                damageTypes,
+            };
+        });
+    }
+
+    /** Get a elemental-blast configuration, throwing an error if none is found according to the arguments passed. */
+    #getBlastConfig(element: ElementTrait, damageType: DamageType): ElementalBlastConfig {
+        const config = this.configs.find(
+            (c) => c.element === element && c.damageTypes.some((t) => t.value === damageType)
+        );
+        if (!config) {
+            throw ErrorPF2e(
+                `Elemental blast configuration of element ${element} and damage type ${damageType} not found.`
+            );
+        }
+
+        return config;
+    }
+
+    /** Get all elemental-blast traits applicable to an element and damage-type combination. */
+    #getTraits(element: ElementTrait, damageType: DamageType): ActionTrait[] {
+        const baseTraits = this.item?.system.traits.value ?? [];
+        return R.uniq(
+            [...baseTraits, element, damageType].filter((t): t is ActionTrait => t in CONFIG.PF2E.actionTraits)
+        ).sort();
+    }
+
+    /** Make an impulse attack roll as part of an elemental blast. */
+    async attack(params: BlastAttackParams): Promise<Rolled<CheckRoll> | null> {
+        const { item, statistic, actionCost } = this;
+        const actionSlug = "elemental-blast";
+        const kineticAura = this.actor.itemTypes.effect.find((e) => e.slug === "effect-kinetic-aura");
+        if (!(statistic && item)) throw ErrorPF2e("Unable to blast");
+        if (!kineticAura) throw ErrorPF2e("No kinetic gate");
```

Comment:
Is the blast rendering on the sheet conditional to the gate being open? If not this should be a UI warning/information for the player to Channel Elements first.